### ISSUE TITLE
Allow program to run after specific amount of idle time

### ISSUE THREAD

**Issue by OlaoluwaM** [Eugeny/tabby#5482]
**Is your feature request related to a problem? Please describe.**
No, my feature request is not related to any problems

**Describe the solution you'd like**
A setting that allows for the configuration of what command to run after a specific amount of idle time (the amount of time where the terminal has not been in use) has passed. Alongside an option for setting the command to be run, like `cmatrix` or any cool/useful effects

**Describe alternatives you've considered**
Using a terminal multiplexer like tmux

**Additional context**
To be frank, I am not even sure if this is a good idea since tabby is already a bit memory-intensive 😅

**Comment by OlaoluwaM** [comment]
With a bit of guidance, I can work on this!!

**Comment by Eugeny** [comment]
Sure, you're looking for the getting started guide: https://github.com/Eugeny/tabby/blob/master/HACKING.md - and then to [watch for newly added tabs](https://docs.tabby.sh/classes/AppService.html#tabOpened_) and listen for their [`activity` event](https://docs.tabby.sh/classes/BaseTabComponent.html#activity_)

**Comment by OlaoluwaM** [comment]
![image](https://user-images.githubusercontent.com/37044906/149821921-e577e9f4-54b9-41de-bc76-6ae3f56999e9.png)
I am getting this issue when I try to build. Also, is it fine to use `npm` for this project?

**Comment by OlaoluwaM** [comment]
I run Linux btw, Fedora 35

**Comment by Eugeny** [comment]
Have you run `yarn` first? No, it's not fine to use `npm` here because specific dependencies are pinned in `yarn.lock`.

**Comment by OlaoluwaM** [comment]
Still getting the same error, and yes I ran `yarn`

**Comment by OlaoluwaM** [comment]
I see in the `HACKING.md` there are some packages to be installed for the Debian and Ubuntu distributions. Is this the case for other Linux distros like Fedora?

**Comment by OlaoluwaM** [comment]
I keep getting errors when building and running tabby locally on Fedora 35. Am I missing something?

**Comment by Eugeny** [comment]
Does `app/node_modules/rxjs` exist? Yes, for Fedora, you need to install the corresponding Fedora packages, but that's not related to your current error.

**Comment by OlaoluwaM** [comment]
Yes!

**Comment by OlaoluwaM** [comment]
Oh, it worked. I didn't know we had to run `yarn` in both the root and `app` folder

**Comment by OlaoluwaM** [comment]
This what I got when invoking `yarn start`
![image](https://user-images.githubusercontent.com/37044906/149907356-5aebe489-6144-4303-9b46-3b34971a08c3.png)

**Comment by Eugeny** [comment]
Sorry! I've just found and fixed an error in `install-deps.js` that prevented it from running automatically. Now you just need to run `node ./scripts/build-native.js` to build the native modules.

**Comment by OlaoluwaM** [comment]
No worries!

**Comment by OlaoluwaM** [comment]
Now I am getting this
![image](https://user-images.githubusercontent.com/37044906/149935817-4db9a375-2bbc-4e62-b420-266fd7bb566b.png)

**Comment by OlaoluwaM** [comment]
This is what I get when I try to run `./scripts/build-native.js`
![image](https://user-images.githubusercontent.com/37044906/149938626-0d345a70-4b85-465c-bc14-df73309db236.png)

**Comment by Eugeny** [comment]
You need to install the normal C++ toolchain (probably https://developer.fedoraproject.org/tech/languages/c/cpp_installation.html)

Plugins must be Angular modules, yes.

**Comment by OlaoluwaM** [comment]
And now I get this when I run `./scripts/build-native.js`
![image](https://user-images.githubusercontent.com/37044906/149953512-32db883d-b0c9-41d5-adf0-d3c7240fb674.png)
![image](https://user-images.githubusercontent.com/37044906/149953750-94ea71d8-e2f5-4ae0-ba7b-6771a54d4b5c.png)

**Comment by Eugeny** [comment]
Have you installed `libfontconfig-dev`? It's one of the requirements listed in https://github.com/Eugeny/tabby/blob/master/HACKING.md

**Comment by OlaoluwaM** [comment]
It's not in the fedora repository
![image](https://user-images.githubusercontent.com/37044906/149954925-cd0e44a1-f69c-4842-b430-46d1095fe0b1.png)

**Comment by Eugeny** [comment]
It is, and if it weren't it doesn't mean you can just skip it: https://fedora.pkgs.org/35/fedora-x86_64/fontconfig-devel-2.13.94-3.fc35.x86_64.rpm.html

**Comment by OlaoluwaM** [comment]
Shoot, my bad seems like lib is a prefix for debian based distros, thanks for letting me know

**Comment by OlaoluwaM** [comment]
Couldn't find the following
 - (libsecret-1-dev) secret-1-dev:  seems to only be available on Debian/Ubuntu
 - (libarchive-tools) archive-tools: Not available either
 - (libnss3) nss3: Not found in Fedora
 - (libatk1.0-0) atk1.0-0: Not found
 - (libatk-bridge2.0-0) atk-bridge2.0-0: Nope
 - (libgdk-pixbuf2.0-0) gdk-pixbuf2.0-0: Nada
 - (libgtk-3-0) gtk-3-0: Nothing
 - (libgbm1) gbm1: Nothing either

**Comment by OlaoluwaM** [comment]
I checked both the https://fedora.pkgs.org/35 link and using the `dnf search` command. I still get this when I run `./scripts/build-native.js`
![image](https://user-images.githubusercontent.com/37044906/149959725-5db6f1e1-e5a0-4d09-b404-533c1b004879.png)

**Comment by OlaoluwaM** [comment]
Am I missing something?

**Comment by Eugeny** [comment]
No, these aren't errors. If you're getting exitcode 0 at the end, you're good to go

**Comment by OlaoluwaM** [comment]
Hmm, so I got it working, but I don't seem to have `tabby` as a command

**Comment by OlaoluwaM** [comment]
![image](https://user-images.githubusercontent.com/37044906/149979352-00027582-0521-4897-b5b6-93ccebdbe264.png)

**Comment by OlaoluwaM** [comment]
I did install it using the `.rpm` on the release page

**Comment by Eugeny** [comment]
True! Looks like it's missing - I'll look into it. You can use the full path(`/opt/Tabby/tabby` instead)

**Comment by OlaoluwaM** [comment]
Also, running `yarn` in the plugin template directory errors out with a status code of 1, because of `node-sass` This is part of the output. It's long. Is there a specific node version to use?
![image](https://user-images.githubusercontent.com/37044906/149986084-1dc0b41b-34bf-45de-826f-bc65db4e47df.png)

**Comment by OlaoluwaM** [comment]
Can I update the deps of the `tabby-clippy` plugin to the latest?

**Comment by OlaoluwaM** [comment]
It's working but when I run `TABBY_PLUGINS=$(pwd) /opt/Tabby/tabby --debug` nothing happens. How do I debug things?

**Comment by OlaoluwaM** [comment]
How do you use the `AppService`. Sorry, I am new to angular (a react dev)

![image](https://user-images.githubusercontent.com/37044906/149997435-d5f77dce-a572-402b-aae3-51f8d0baab6d.png)

**Comment by Eugeny** [comment]
You just add them as constructor parameters: https://angular.io/guide/architecture-services#dependency-injection-di

**Comment by OlaoluwaM** [comment]
Thank you! So I got it working to the point where we are able to log out to the console when the active tab changes. However, I am a bit stuck. What I am trying to implement is the ability for a command to be run after a tab has been idle (no command is running on the foreground) for a specific amount of time, with said time being a user-specified input. Currently, there doesn't seem to be a way to get the amount of time that has passed while the tab has been idle, but there is a way to get the current process (assuming it only returns what's running on the foreground and not the background)

**Comment by OlaoluwaM** [comment]
It requires a bit of polymorphism, but I am not familiar with how angular handles that or how to approach it in this plugin. Nevertheless, my thought for an implementation is to have an interval for the user-specified duration. Once the interval ends, check if the currently active tab is still idle (no foreground processes). If it is, clean up the interval and invoke whatever command the user-specified. When the tab is no longer idle, stop the command and revert things back to normal. If a user switches tabs for some reason, the plugin could either keep the command running on the previously active tab, or stop it in the previous tab and restart the idle-checking process on the new active tab. I think the latter may be better for resource utilization

**Comment by Eugeny** [comment]
Just use the `activity$` event and a bit of RxJS: 
```js
tab.activity$.pipe(
debounceTime(10000)).subscribe(() => console.log('no activity for 10s'))
```

**Comment by OlaoluwaM** [comment]
> ```js
> tab.activity$.pipe(
> 	filter(x => x),
> 	debounceTime(10000)
> ).subscribe(() => console.log('no activity for 10s'))
> ```

Or I could do this... 😅

**Comment by OlaoluwaM** [comment]
> Just use the `activity$` event and a bit of RxJS:
> 
> ```js
> tab.activity$.pipe(
> 	filter(x => x),
> 	debounceTime(10000)
> ).subscribe(() => console.log('no activity for 10s'))
> ```

So I tried this but it doesn't work when there is a process running on the terminal like `termdown`. Meaning, even if termdown is running, after 10s "no activity for 10s" is logged :/

**Comment by Eugeny** [comment]
Well, yes, `activity$` refers to output being printed. You could check `tab.getCurrentProcess()` whether there's still a process running.

**Comment by OlaoluwaM** [comment]
Hmm, looks like `getCurrentProcess()` keeps returning null. I checked the source and it looks like I will need to override it. How can I do that?

**Comment by Eugeny** [comment]
That's weird, does the current process show up in the tab context menu? (it also uses `getCurrentProcess()`). Also make sure you're actually checking for the current process when clicking the tab and not then attaching to the tab.

**Comment by lppsuixn** [comment]
@OlaoluwaM hi,how is it going? I need that! thx
